src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -805,10 +810,14 @@ void Connection::handle_connect_wait(Status status)
 
     REALM_ASSERT_EX(m_state == ConnectionState::connecting, m_state);
     logger.info("Connect timeout"); // Throws
-    involuntary_disconnect(
-        SessionErrorInfo{Status{ErrorCodes::SyncConnectTimeout, "Sync connection was not fully established in time"},
-                         IsFatal{false}},
-        ConnectionTerminationReason::sync_connect_timeout); // Throws
+    SessionErrorInfo error_info({ErrorCodes::SyncConnectTimeout, "Sync connection was not fully established in time"},
+                                IsFatal{false});
+    // If the connection fails/times out and the server has not been contacted yet, refresh the location
+    // to make sure the websocket URL is correct
+    if (!m_server_endpoint.is_verified) {
```

Comment:
could we end up in a loop here? or does the retry strategy kick in?

Reply:
Actually it shouldn't, since with this failure, it will kick off an access token refresh with location update, which will just enter the reconnect backoff wait if that operation fails. The session will continue the sequence of connect to websocket (fails), request location (fails) and wait backoff delay until the client goes online or the session is closed.

Reply:
so it is some sort of loop (which is fine), and as long as the backoff kicks in that's okay

Reply:
The sequence looks like this (from the updated `"Sync Session retries after initial location failure"` test):
```
Realm.Sync.Client.Session - Connection[1]: Connecting to 'ws://alternate.someurl.fake:80/api/client/v2.0/app/fake-app-id/realm-sync'
Realm - State: start
Realm.Sync.Client.Session - Connection[1]: Closing the websocket with error code=WebSocket: Connection Failed, message='404 not found', was_clean=false
Realm.App - App: refresh_access_token: email: Ursus ursinus (updating location)
Realm.App - App: request location: http://alternate.someurl.fake/api/client/v2.0/app/fake-app-id/location
Realm.App - ERROR: App: request location failed (HTTPError 404): http error code considered fatal: 404 not found. Client Error: 404
Realm.Sync.Client.Session - Connection[1]: Allowing reconnection in 956 milliseconds
Realm.Sync.Client.Session - Connection[1]: Connecting to 'ws://alternate.someurl.fake:80/api/client/v2.0/app/fake-app-id/realm-sync'
Realm - State: start
Realm.Sync.Client.Session - Connection[1]: Closing the websocket with error code=WebSocket: Connection Failed, message='404 not found', was_clean=false
Realm.App - App: refresh_access_token: email: Ursus ursinus (updating location)
Realm.App - App: request location: http://alternate.someurl.fake/api/client/v2.0/app/fake-app-id/location
Realm.App - ERROR: App: request location failed (HTTPError 404): http error code considered fatal: 404 not found. Client Error: 404
Realm.Sync.Client.Session - Connection[1]: Allowing reconnection in 1738 milliseconds
Realm.Sync.Client.Session - Connection[1]: Connecting to 'ws://alternate.someurl.fake:80/api/client/v2.0/app/fake-app-id/realm-sync'
Realm - State: start
Realm.Sync.Client.Session - Connection[1]: Closing the websocket with error code=WebSocket: Connection Failed, message='404 not found', was_clean=false
Realm.App - App: refresh_access_token: email: Ursus ursinus (updating location)
Realm.App - App: request location: http://alternate.someurl.fake/api/client/v2.0/app/fake-app-id/location
Realm.App - App: request location: http://redirect.someurl.fake:8081/api/client/v2.0/app/fake-app-id/location
Realm.App - App: Location info returned for deployment model: GLOBAL(US-VA)
Realm.App - App: update_hostname: http://redirect.someurl.fake:8081 | ws://redirect.someurl.fake:8081 | base URL: http://alternate.someurl.fake
Realm.Sync.Client.Session - Connection[1]: Allowing reconnection in 3509 milliseconds
```

Reply:
why is the session restarted? (is it driven by the test?) it still gets a 404.

Reply:
When the ws host url is updated via querying the location, the session is restarted since, most likely, this url value has changed.
FYI - the test doesn't have a "real" socket provider (it's a non-baas test), so it is hard coded to always return 404.